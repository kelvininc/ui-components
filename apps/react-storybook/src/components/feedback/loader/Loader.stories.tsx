import type { Meta, StoryObj } from '@storybook/react';
import { KvLoader } from '@kelvininc/react-ui-components';

const meta = {
	title: 'Feedback/Loader',
	component: KvLoader,
	argTypes: {
		isLoading: { type: 'boolean' },
		hasOverlay: { type: 'boolean' }
	}
} satisfies Meta<typeof KvLoader>;

export default meta;
type Story = StoryObj<typeof meta>;

export const IsLoading: Story = {
	args: {
		isLoading: true,
		hasOverlay: false
	}
};

export const HasOverlay: Story = {
	args: {
		isLoading: true,
		hasOverlay: true
	}
};
