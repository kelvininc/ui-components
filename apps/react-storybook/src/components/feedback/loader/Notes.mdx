import Readme from "@kelvininc/ui-components/docs/components/loader/readme.md?raw";

import { Markdown, Meta, ArgTypes } from "@storybook/blocks";
import * as LoaderStories from "./Loader.stories";

<Meta of={LoaderStories} />
<Markdown>{Readme}</Markdown>
