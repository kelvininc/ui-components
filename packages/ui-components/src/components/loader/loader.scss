@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --loader-overlay-color: Overlay color
	 * @prop --loader-overlay-opacity: Overlay opacity
	 * @prop --loader-background-color: Loader background
	 * @prop --loader-spinner-color: Loader color
	 * @prop --loader-z-index: Loader z-index position.
	 */
	--loader-overlay-color: #{kv-color('neutral-4')};
	--loader-overlay-opacity: 0.4;
	--loader-background-color: #{kv-color('neutral-2')};
	--loader-spinner-color: #{kv-color('secondary-5')};
	--loader-padding: #{15px};
	--loader-border-size: #{6px};
	--loader-z-index: 1;
}

.loader-container {
	width: 100%;
	height: 100%;
	position: absolute;
	top: 0;
	left: 0;
	z-index: var(--loader-z-index);

	.overlay {
		width: 100%;
		height: 100%;
		opacity: var(--loader-overlay-opacity);
		background-color: var(--loader-overlay-color);
	}

	.loader {
		height: 0;
		width: 0;
		padding: var(--loader-padding);
		border: var(--loader-border-size) solid var(--loader-background-color);
		border-right-color: var(--loader-spinner-color);
		border-radius: 50%;
		animation: loadrotate 1s infinite linear;
		position: absolute;
		top: 0;
		left: 0;
		bottom: 0;
		right: 0;
		margin: auto;
	}
}

@keyframes loadrotate {
	/* Use 0% instead for anticlockwise */
	100% {
		transform: rotate(360deg);
	}
}
