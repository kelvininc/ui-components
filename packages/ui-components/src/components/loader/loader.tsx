import { Component, Host, Prop, h } from '@stencil/core';

@Component({
	tag: 'kv-loader',
	styleUrl: 'loader.scss',
	shadow: true
})
export class KvLoader {
	/** (optional) If `true` the loader is enabled */
	@Prop({ reflect: true }) isLoading: boolean = false;
	/** (optional) If `true` the loader is overlay */
	@Prop({ reflect: true }) hasOverlay: boolean = false;

	render() {
		return (
			<Host>
				{this.isLoading && (
					<div class="loader-container">
						{this.hasOverlay && <div class="overlay"></div>}
						<div class="loader"></div>
					</div>
				)}
			</Host>
		);
	}
}
