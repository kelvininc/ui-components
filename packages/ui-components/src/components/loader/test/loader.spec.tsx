import { SpecPage } from '@stencil/core/internal';
import { KvLoader } from '../loader';
import { newSpecPage } from '@stencil/core/testing';

describe('Loader (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvLoader],
				html: `<kv-loader></kv-loader>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when is loading', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvLoader],
				html: `<kv-loader is-loading></kv-loader>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has overlay', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvLoader],
				html: `<kv-switch-button is-loading has-overlay></kv-switch-button>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
