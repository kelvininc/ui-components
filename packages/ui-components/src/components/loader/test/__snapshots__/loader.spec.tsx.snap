// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader (unit tests) when has overlay should match the snapshot 1`] = `
<kv-switch-button
  has-overlay=""
  is-loading=""
/>
`;

exports[`Loader (unit tests) when is loading should match the snapshot 1`] = `
<kv-loader is-loading="">
  <template shadowrootmode="open">
    <div class="loader-container">
      <div class="loader"></div>
    </div>
  </template>
</kv-loader>
`;

exports[`Loader (unit tests) when uses default props should match the snapshot 1`] = `
<kv-loader>
  <template shadowrootmode="open"></template>
</kv-loader>
`;
